Ultra ram nov22

amartens opened the pull request:
pfb_fir and fft_wideband_real Ultra RAM option. 

PFB FIR includes delays and coefficients.
FFT Wideband Real only supports delays for now.

andrewvanderbyl committed:
Merge branch 'm2021a' of github.com:ska-sa/mlib_devel into m2021a

andrewvanderbyl committed:
Added changes to bus_dual_port_ram_init.m to accommodate Ultra RAM.
Added platfor_devices to assist in platform lookup (to see if Ultra RAM supported).

andrewvanderbyl committed:
Added UltraRAM option to pfb_fir_taps

andrewvanderbyl committed:
Added UltraRAM option to pfb_fir_generic.

andrewvanderbyl committed:
Added Ultra RAM option to pfb_fir_cieff_gen.

andrewvanderbyl committed:
Commit for casper_library.slx

andrewvanderbyl committed:
Changed mechanism to check if Ultra RAM available based on XSG block.
Set memory to default to Block RAM if Ultra RAM selected but platform
selected doesn't support this.

NOTE: This is only applied to coeff generator. Top level coeff gen block
does not yet change to Block RAM if defalut mem_type changed due to
non-support of selected platform. pfb_fir_taps needs to be updated.

andrewvanderbyl committed:
Added rollback to default (Block RAM) for top level block as well.
This all needs to be retested.

andrewvanderbyl committed:
Fixed bug for UltraRAM selection if dropping in a new fir_generic block.
All seems to work.

andrewvanderbyl committed:
Merge branch 'UltraRAM_Nov22' of github.com:ska-sa/mlib_devel into UltraRAM_Nov22

amartens committed:
fixed bus_dual_port_ram block in bus library

Rolled bus library back.
Copied out block. Removed library link.
Changed parameter to have Ultra RAM as an option.
Removed old block from library and copied in new.
Saved library.
Tested pfb_fir_generic works with the new block as required.

amartens committed:
split buffer implementation into taps and coeffs

Data taps and coefficient storage implementation are now separate
options.
Distributed memory removed as an option as scripts can't draw these
properly as the ports change. (This could be added with a little effort
but would only make sense for small PFBs).
Removed the checks of the implementation target at a high level as
checks already done at a low level.

amartens committed:
Added memory option for delay_bram

Block and init script to allow Ultra RAM buffer to be used.

amartens committed:
fft_stage_n_init option for delays in Ultra RAM

Block parameter option added and init script support.

amartens committed:
URAM delays option in biplex_core

A number of delay buffers in stages can now be implemented in URAM.

amartens committed:
UltraRAM delay option for fft_biplex_real_4x

Passed down to internal biplex_core.

amartens committed:
fft_wideband_real Ultra RAM delay option

Delays can now be implemented in Ultra RAM for a certain number of
stages.

The "FFT Stages" parameter default value of -1 was changed to 0 as
the script was not able to execute correctly with -1.

andrewvanderbyl reviewed (changes requested):
FIR seems to behave as expected (both BRAM and UltraRAM options), but I get errors if dropping in a fft_wide_band_real block. Problem seems to be with the fft_biplex_real_4x block. See attached images.
![fft_wide_band_real_error](https://user
![fft_wide_band_real_error](https://user-images.githubusercontent.com/11848124/216539666-8de56a78-16a2-4fd9-8f39-81588844e850.png)
-images.githubusercontent.com/11848124/216539625-3b0e281f-0c2f-4320-afe5-c7fcb2c77a59.png)

amartens committed:
Merge branch 'm2021a' of github.com:ska-sa/mlib_devel into UltraRAM_Nov22

amartens committed:
Added Ultra RAM as option to bus_single_port_ram

amartens committed:
Removed Ultra RAM as option for coefficients

Ultra RAM does not permit initialisaion values so cannot be used dto
store coefficients. Removed this options from the PFB FIR.

amartens commented:
Oh dear. I hope that is it now.

andrewvanderbyl reviewed (approved):
Looks good. Seems to work.
